packages/parse5-sax-parser/lib/index.ts

```diff
@@ -0,0 +1,312 @@
+
+        return true;
+    }
+
+    protected _emitToken(eventName: string, token: SaxToken): void {
+        this.emit(eventName, token);
+    }
+
+    private _emitPendingText(): void {
+        if (this.pendingText !== null) {
+            this._handleToken(this.pendingText);
+            this.pendingText = null;
+        }
+    }
+}
+
+export interface SaxToken {
+    /** Source code location info. Available if location info is enabled via {@link SAXParserOptions}. */
+    sourceCodeLocation?: Location | null;
+}
+
+export interface StartTag extends SaxToken {
+    /** Tag name */
+    tagName: string;
+    /** List of attributes */
+    attrs: Attribute[];
+    /** Indicates if the tag is self-closing */
+    selfClosing: boolean;
+}
+
+export interface EndTag extends SaxToken {
+    /** Tag name */
+    tagName: string;
+}
+
+export interface Text extends SaxToken {
+    /** Text content. */
+    text: string;
+}
+
+export interface Comment extends SaxToken {
+    /** Comment text. */
+    text: string;
+}
+
+export interface Doctype extends SaxToken {
+    /** Document type name. */
+    name: string | null;
+    /** Document type public identifier. */
+    publicId: string | null;
+    /** Document type system identifier. */
+    systemId: string | null;
+}
+
+const TEXT_EMISSION_HELPER = {
+    eventName: 'text',
+    reshapeToken: (origToken: CharacterToken): Text => ({
+        text: origToken.chars,
+        sourceCodeLocation: origToken.location,
+    }),
+};
+
+const TOKEN_EMISSION_HELPERS = {
+    [TokenType.START_TAG]: {
+        eventName: 'startTag',
+        reshapeToken: (origToken: TagToken): StartTag => ({
+            tagName: origToken.tagName,
+            attrs: origToken.attrs,
+            selfClosing: origToken.selfClosing,
+            sourceCodeLocation: origToken.location,
+        }),
+    },
+    [TokenType.END_TAG]: {
+        eventName: 'endTag',
+        reshapeToken: (origToken: TagToken): EndTag => ({
+            tagName: origToken.tagName,
+            sourceCodeLocation: origToken.location,
+        }),
+    },
+    [TokenType.COMMENT]: {
+        eventName: 'comment',
+        reshapeToken: (origToken: CommentToken): Comment => ({
+            text: origToken.data,
+            sourceCodeLocation: origToken.location,
+        }),
+    },
+    [TokenType.DOCTYPE]: {
+        eventName: 'doctype',
+        reshapeToken: (origToken: DoctypeToken): Doctype => ({
+            name: origToken.name,
+            publicId: origToken.publicId,
+            systemId: origToken.systemId,
+            sourceCodeLocation: origToken.location,
+        }),
+    },
+    [TokenType.CHARACTER]: TEXT_EMISSION_HELPER,
+    [TokenType.NULL_CHARACTER]: TEXT_EMISSION_HELPER,
+    [TokenType.WHITESPACE_CHARACTER]: TEXT_EMISSION_HELPER,
+    [TokenType.HIBERNATION]: {
+        eventName: 'hibernation',
+        reshapeToken: (): Record<string, never> => ({}),
+    },
+};
+
+export interface SAXParser {
```

Comment:
just so i understand, is this to cause typescript to merge the declarations and provide extra type information on top of the base class' `on` signature? i.e. this'll merge the interface here and the class above, so you don't have to override `on` to define the extra signatures.

Reply:
and if that's the case, do you really need the base one at the bottom? won't that be inherited still?

Reply:
Your understanding is spot on, TS will merge these declarations and provide types for the events (see https://github.com/inikulin/parse5/runs/4775155629?check_suite_focus=true for an example of an error this caused when I had two token types swapped).

The order doesn't matter, happy to change it if there is a strong preference either way.

Reply:
what i mean is if you left the base one out completely, it'd still pick it up from the base class, wouldn't it? maybe we can drop it and keep the rest?

Reply:
Not sure I follow. Could you elaborate?

Reply:
There will be an error if the last one is not provided, even if it is defined in the base type 😕

Reply:
well that sucks! 😂  so you can't specify overloads without also re-specifying the base signature? sad times, typescript